src/coreclr/jit/hwintrinsicxarch.cpp

```diff
@@ -1459,31 +1527,33 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
         case NI_Vector512_ConvertToSingle:
         {
             assert(sig->numArgs == 1);
-
+            assert(varTypeIsInt(simdBaseType));
+            intrinsic = NI_Illegal;
             if (simdBaseType == TYP_INT)
             {
-                switch (simdSize)
+                if (simdSize == 16)
                 {
-                    case 16:
-                        intrinsic = NI_SSE2_ConvertToVector128Single;
-                        break;
-                    case 32:
-                        intrinsic = NI_AVX_ConvertToVector256Single;
-                        break;
-                    case 64:
-                        intrinsic = NI_AVX512F_ConvertToVector512Single;
-                        break;
-                    default:
-                        unreached();
+                    intrinsic = NI_SSE2_ConvertToVector128Single;
+                }
+                else if (simdSize == 32 && compOpportunisticallyDependsOn(InstructionSet_AVX))
+                {
+                    intrinsic = NI_AVX_ConvertToVector256Single;
+                }
+                else if (simdSize == 64 && IsBaselineVector512IsaSupportedOpportunistically())
+                {
+                    intrinsic = NI_AVX512F_ConvertToVector512Single;
                 }
-
-                op1     = impSIMDPopStack();
-                retNode = gtNewSimdHWIntrinsicNode(retType, op1, intrinsic, simdBaseJitType, simdSize);
```

Comment:
Same general question, it's not clear why this path had to change at all.

`simdSize == 32` and `simdSize == 64` should have already been filtered out if `Avx` or `Avx512F` weren't supported, so the checks there are redundant.

It seems like we could've just updated `retNode = gtNewSimd...` here either by having one here and one in the `TYP_UINT` path or by moving just it down under the `if (intrinsic != NI_Illegal)` without changing the switch.

Reply:
I wasn't taking into account that if ```simdSize==32``` will be filtered out if AVX wasn't supported and same for ```simdSize==64``` & AVX512F. Moved them to what is was and just added a ```switch``` for ```TYP_UINT```.